storage/badger/protocol_kv_store.go

```diff
@@ -0,0 +1,172 @@
+			}
+			return &kvStore, nil
+		}
+	}
+	storeByStateID := func(stateID flow.Identifier, data *storage.KeyValueStoreData) func(*transaction.Tx) error {
+		return transaction.WithTx(operation.InsertProtocolKVStore(stateID, data))
+	}
+
+	storeByBlockID := func(blockID flow.Identifier, stateID flow.Identifier) func(*transaction.Tx) error {
+		return func(tx *transaction.Tx) error {
+			err := transaction.WithTx(operation.IndexProtocolKVStore(blockID, stateID))(tx)
+			if err != nil {
+				return fmt.Errorf("could not index protocol state for block (%x): %w", blockID[:], err)
+			}
+			return nil
+		}
+	}
+
+	retrieveByBlockID := func(blockID flow.Identifier) func(tx *badger.Txn) (flow.Identifier, error) {
+		return func(tx *badger.Txn) (flow.Identifier, error) {
+			var stateID flow.Identifier
+			err := operation.LookupProtocolState(blockID, &stateID)(tx)
+			if err != nil {
+				return flow.ZeroID, fmt.Errorf("could not lookup protocol state ID for block (%x): %w", blockID[:], err)
+			}
+			return stateID, nil
+		}
+	}
+
+	return &ProtocolKVStore{
+		db: db,
+		cache: newCache[flow.Identifier, *storage.KeyValueStoreData](collector, metrics.ResourceProtocolKVStore,
+			withLimit[flow.Identifier, *storage.KeyValueStoreData](stateCacheSize),
+			withStore(storeByStateID),
+			withRetrieve(retrieveByStateID)),
+		byBlockIdCache: newCache[flow.Identifier, flow.Identifier](collector, metrics.ResourceProtocolKVStoreByBlockID,
+			withLimit[flow.Identifier, flow.Identifier](stateByBlockIDCacheSize),
+			withStore(storeByBlockID),
+			withRetrieve(retrieveByBlockID)),
+	}
+}
+
+// StoreTx returns an anonymous function (intended to be executed as part of a badger transaction),
+// which persists the given model as part of a DB tx.
+// Expected errors of the returned anonymous function:
+//   - storage.ErrAlreadyExists if a model with the given id is already stored
+func (s *ProtocolKVStore) StoreTx(stateID flow.Identifier, data *storage.KeyValueStoreData) func(*transaction.Tx) error {
+	return s.cache.PutTx(stateID, data)
+}
+
+// IndexTx returns an anonymous function intended to be executed as part of a database transaction.
+// In a nutshell, we want to maintain a map from `blockID` to `stateID`, where `blockID` references the
+// block that _proposes_ updated key-value store.
+// Upon call, the anonymous function persists the specific map entry in the node's database.
+// Protocol convention:
+//   - Consider block B, whose ingestion might potentially lead to an updated KV store. For example,
+//     the KV store changes if we seal some execution results emitting specific service events.
+//   - For the key `blockID`, we use the identity of block B which _proposes_ this updated KV store. As value,
+//     the hash of the resulting key-value store state at the end of processing B is to be used.
+//   - CAUTION: The updated state requires confirmation by a QC and will only become active at the child block,
+//     _after_ validating the QC.
```

Comment:
We make the same statement in the storage abstraction for the protocol state:
https://github.com/onflow/flow-go/blob/81266cf708ce648e14d871cace67563b504ba42c/storage/protocol_state.go#L27-L28

For sure, this is a simplified statement skipping over some details. Nevertheless, I feel the statement is still applicable, and a similarly worded statement is also in our notion doc [`Identity-Changing Operations`](https://www.notion.so/dapperlabs/Identity-Changing-Operations-610bb2d3b63f4d3dae12e405d3c5a097?pvs=4):
> <img width="718" alt="Screenshot 2024-02-14 at 12 12 59 PM" src="https://github.com/onflow/flow-go/assets/26322303/4bdf95b4-04ec-4f0e-99dd-a353cc6478a4">

My interpretation of your comment is: "the consensus nodes don't need to wait for a QC, as they validate the correctness of the state transition by re-computing it locally" I am happy to discuss how we could reword the sentence, to make it more clear. 

Nevertheless, I feel there is a super-important aspect in here:
> **The updated state** [...] **only become active at the child block**, **after** [...] **valid QC**


This avoids some subtle but safety-critical misusage:
1. using the resulting protocol state (kv-store) and apply it to the payload of the block (circular dependency) 
2. some off-chain logic uses the protocol state to determine signing authority. Without waiting for a valid QC, byzantine  consensus participant could give a malicious block to a client where it tempered with the epoch participation information. Thereby, a client could be easily tricked into accepting malicious messages from unauthorized parties, or rejecting messages from honest parties. 

On the one hand, the consensus follower handles this safely and protects higher-level business logic. On the other hand, I feel we should emphasize this safety-critical requirement because not all applications might be using the consensus follower and additionally we want to be reminded of this requirement when we work on the consensus follower in the future.

Reply:
how about we write (?)
```suggestion
	//   - CAUTION: The protocol state will only become active at the child block, _after_ its correctness has
	//     been affirmed by a valid QC.
```
Does that address your wording concerns?